libnavigation-core/src/test/java/com/mapbox/navigation/core/replay/ReplayLocationEngineTest.kt

```diff
@@ -0,0 +1,228 @@
+        val successData = mutableListOf<LocationEngineResult>()
+        val failureData = slot<Exception>()
+        val callback: LocationEngineCallback<LocationEngineResult> = mockk {
+            every { onSuccess(capture(successData)) } returns Unit
+            every { onFailure(capture(failureData)) } returns Unit
+        }
+
+        replayLocationEngine.requestLocationUpdates(mockk(), callback, mockk())
+        replayLocationEngine.replayEvents(listOf(testLocation()))
+        replayLocationEngine.replayEvents(listOf(testLocation()))
+
+        assertEquals(2, successData.size)
+        assertFalse(failureData.isCaptured)
+    }
+
+    @Test
+    fun `removeLocationUpdates should stop capturing locations`() {
+        val successData = mutableListOf<LocationEngineResult>()
+        val failureData = slot<Exception>()
+        val callback: LocationEngineCallback<LocationEngineResult> = mockk {
+            every { onSuccess(capture(successData)) } returns Unit
+            every { onFailure(capture(failureData)) } returns Unit
+        }
+
+        replayLocationEngine.requestLocationUpdates(mockk(), callback, mockk())
+        replayLocationEngine.replayEvents(listOf(testLocation()))
+        replayLocationEngine.removeLocationUpdates(callback)
+        replayLocationEngine.replayEvents(listOf(testLocation()))
+
+        assertEquals(1, successData.size)
+        assertFalse(failureData.isCaptured)
+    }
+
+    @Test
+    fun `should map all values for a location`() {
+        val successData = slot<LocationEngineResult>()
+        val callback: LocationEngineCallback<LocationEngineResult> = mockk {
+            every { onSuccess(capture(successData)) } returns Unit
+        }
+
+        replayLocationEngine.requestLocationUpdates(mockk(), callback, mockk())
+        replayLocationEngine.replayEvents(
+            listOf(
+                ReplayEventUpdateLocation(
+                    1234.56,
+                    ReplayEventLocation(
+                        lat = 38.571011,
+                        lon = -121.452869,
+                        provider = "ReplayLocationEngineTest",
+                        time = 2468.135,
+                        altitude = 158.32,
+                        accuracyHorizontal = 12.3,
+                        bearing = 287.4,
+                        speed = 12.2
+                    )
+                )
+            )
+        )
+
+        val lastLocation = successData.captured.lastLocation!!
+        assertEquals(38.571011, lastLocation.latitude, 0.000001)
+        assertEquals(-121.452869, lastLocation.longitude, 0.000001)
+        assertEquals("ReplayLocationEngineTest", lastLocation.provider)
+        assertEquals(158.32, lastLocation.altitude, 0.01)
+        assertEquals(12.3f, lastLocation.accuracy, 0.01f)
+        assertEquals(287.4f, lastLocation.bearing, 0.01f)
+        assertEquals(12.2f, lastLocation.speed, 0.01f)
+    }
+
+    @Test
+    fun `should map monotonic time`() {
```

Comment:
left out the System.currentTimeMillis() test because https://github.com/mockk/mockk/issues/98

Worth considering `Calendar.getInstance().timeInMillis` instead of `Date()`
But it becomes more complicated with timezones

My preferred approach for testing time, is to create a `MapboxTime` interface that can be mocked. there's a couple open source options but that would mainly be motivated by incorporating NTP time
- https://github.com/instacart/truetime-android
- https://github.com/lyft/Kronos-Android 

conclusion: leaving that test out because it's a bigger project with less reward right now

Reply:
> My preferred approach for testing time, is to create a `MapboxTime` interface that can be mocked.

Same here.